fairlearn/metrics/_metric_frame.py

```diff
@@ -303,53 +334,39 @@ def __init__(self,
                 raise ValueError(_DUPLICATE_FEATURE_NAME.format(name))
             nameset.add(name)
 
-        self._overall = self._compute_overall(func_dict, y_t, y_p, cf_list)
-        self._by_group = self._compute_by_group(func_dict, y_t, y_p, sf_list, cf_list)
-
-    def _compute_overall(self, func_dict, y_true, y_pred, cf_list):
-        if cf_list is None:
-            result = pd.Series(index=func_dict.keys(), dtype='object')
-            for func_name in func_dict:
-                metric_value = func_dict[func_name].evaluate_all(y_true, y_pred)
-                result[func_name] = metric_value
+        if self._cf_names is None:
+            self._overall = apply_to_dataframe(
+                all_data,
+                metric_functions=annotated_funcs)
         else:
-            result = self._compute_dataframe_from_rows(func_dict, y_true, y_pred, cf_list)
-        return result
+            temp = all_data.groupby(by=self._cf_names).apply(
+                apply_to_dataframe,
+                metric_functions=annotated_funcs
+            )
+            # If there are multiple control features, might have missing combinations
+            if len(self._cf_names) > 1:
+                all_indices = pd.MultiIndex.from_product([x.classes_ for x in cf_list],
+                                                         names=[x.name_ for x in cf_list])
+
+                self._overall = temp.reindex(index=all_indices)
+            else:
+                self._overall = temp
 
-    def _compute_by_group(self, func_dict, y_true, y_pred, sf_list, cf_list):
-        rows = copy.deepcopy(sf_list)
+        grouping_features = copy.deepcopy(sf_list)
         if cf_list is not None:
             # Prepend the conditional features, so they are 'higher'
-            rows = copy.deepcopy(cf_list) + rows
+            grouping_features = copy.deepcopy(cf_list) + grouping_features
 
-        return self._compute_dataframe_from_rows(func_dict, y_true, y_pred, rows)
+        temp = all_data.groupby([x.name_ for x in grouping_features]).apply(
+            apply_to_dataframe,
+            metric_functions=annotated_funcs)
+        if len(grouping_features) > 1:
+            all_indices = pd.MultiIndex.from_product([x.classes_ for x in grouping_features],
```

Comment:
I wouldn't mind if you apply `black` on this file :D

Reply:
Done